token/metadata.go

```diff
@@ -48,3 +53,178 @@ func (m *Metadata) SpentTokenID() []*token2.ID {
+}
+
+// Transfer returns the i-th transfer metadata, if present
+func (m *Metadata) Transfer(i int) (*TransferMetadata, error) {
+	if i >= len(m.tokenRequestMetadata.Transfers) {
+		return nil, errors.Errorf("index [%d] out of range [0:%d]", i, len(m.tokenRequestMetadata.Transfers))
+	}
+	return &TransferMetadata{TransferMetadata: &m.tokenRequestMetadata.Transfers[i]}, nil
+}
+
+// IssueMetadata contains the metadata of an issue action
+type IssueMetadata struct {
+	*api2.IssueMetadata
+}
+
+// Match returns true if the given action matches this metadata
+func (m *IssueMetadata) Match(action *IssueAction) error {
+	if len(m.Outputs) != action.NumOutputs() {
+		return errors.Errorf("expected [%d] outputs but got [%d]", len(m.Outputs), action.NumOutputs())
+	}
+	if len(m.Outputs) != len(m.TokenInfo) {
+		return errors.Errorf("expected [%d] token info but got [%d]", len(m.Outputs), len(m.TokenInfo))
+	}
+	if len(m.Outputs) != len(m.Receivers) {
+		return errors.Errorf("expected [%d] receivers but got [%d]", len(m.Outputs), len(m.Receivers))
+	}
+	if len(m.Outputs) != len(m.ReceiversAuditInfos) {
+		return errors.Errorf("expected [%d] receiver audit infos but got [%d]", len(m.Outputs), len(m.ReceiversAuditInfos))
+	}
+	return nil
+}
+
+// IsOutputAbsent returns true if the given output's metadata is absent
+func (m *IssueMetadata) IsOutputAbsent(j int) bool {
+	return len(m.TokenInfo[j]) == 0
+}
+
+// TransferMetadata contains the metadata of a transfer action
+type TransferMetadata struct {
+	*api2.TransferMetadata
+}
+
+// Match returns true if the given action matches this metadata
+func (m *TransferMetadata) Match(action *TransferAction) error {
+	if len(m.TokenIDs) != 0 && len(m.TokenIDs) != len(m.Senders) {
+		return errors.Errorf("expected [%d] token IDs and senders but got [%d]", len(m.TokenIDs), len(m.Senders))
+	}
+	if len(m.Senders) != len(m.SenderAuditInfos) {
+		return errors.Errorf("expected [%d] senders and sender audit infos but got [%d]", len(m.Senders), len(m.SenderAuditInfos))
+	}
+
+	if len(m.Outputs) != action.NumOutputs() {
+		return errors.Errorf("expected [%d] outputs but got [%d]", len(m.Outputs), action.NumOutputs())
+	}
+	if len(m.Outputs) != len(m.TokenInfo) {
+		return errors.Errorf("expected [%d] token info but got [%d]", len(m.Outputs), len(m.TokenInfo))
+	}
+	if len(m.Outputs) != len(m.Receivers) {
+		return errors.Errorf("expected [%d] receivers but got [%d]", len(m.Outputs), len(m.Receivers))
+	}
+	if len(m.Outputs) != len(m.ReceiverAuditInfos) {
+		return errors.Errorf("expected [%d] receiver audit infos but got [%d]", len(m.Outputs), len(m.ReceiverAuditInfos))
+	}
+	if len(m.Outputs) != len(m.ReceiverIsSender) {
+		return errors.Errorf("expected [%d] receiver is sender but got [%d]", len(m.Outputs), len(m.ReceiverIsSender))
+	}
+	return nil
+}
+
+// IsOutputAbsent returns true if the given output's metadata is absent
+func (m *TransferMetadata) IsOutputAbsent(j int) bool {
```

Comment:
Is this a function we really want to expose? Does the user of `TransferMetadata` even understand how it is structured to provide a useful index?

Reply:
ok, I can see that this is used in `request.go` mhhhh